include/aspect/particle/property/cpo_bingham_average.h

```diff
@@ -0,0 +1,202 @@
+           * @param [in] position The current particle position.
+           *
+           * @param [in] solution The values of the solution variables at the
+           * current particle position.
+           *
+           * @param [in] gradients The gradients of the solution variables at
+           * the current particle position.
+           *
+           * @param [in,out] particle_properties The properties of the particle
+           * that is updated within the call of this function.
+           */
+          virtual
+          void
+          update_one_particle_property (const unsigned int data_position,
+                                        const Point<dim> &position,
+                                        const Vector<double> &solution,
+                                        const std::vector<Tensor<1,dim>> &gradients,
+                                        const ArrayView<double> &particle_properties) const;
+
+          /**
+           * This implementation tells the particle manager that
+           * we need to update particle properties every time step.
+           */
+          UpdateTimeFlags
+          need_update () const;
+
+          /**
+           * Return which data has to be provided to update the property.
+           * The integrated strains needs the gradients of the velocity.
+           */
+          virtual
+          UpdateFlags
+          get_needed_update_flags () const;
+
+          /**
+           * Set up the information about the names and number of components
+           * this property requires.
+           *
+           * @return A vector that contains pairs of the property names and the
+           * number of components this property plugin defines.
+           */
+          virtual
+          std::vector<std::pair<std::string, unsigned int>>
+          get_property_information() const;
+
+          /**
+           * Computes the Bingham average
+           */
+          std::array<std::array<double,3>,3> compute_bingham_average(std::vector<Tensor<2,3>> matrices) const;
+
+          /**
+          * Get volume weighted euler angles, using random draws to convert
+          * to a discrete number of orientations, weighted by volume.
+           */
+          std::vector<Tensor<2,3>>
+          random_draw_volume_weighting(std::vector<double> fv,
+                                       std::vector<Tensor<2,3>> matrices,
+                                       unsigned int n_output_grains) const;
+
+
+          /**
+          * Declare the parameters this class takes through input files.
+          */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int cpo_data_position;
+
+          double rad_to_degree = 180.0/M_PI;
```

Comment:
Let's not start another way to define pi, we already have one:
```suggestion
          constexpr double rad_to_degree = 180.0/numbers::PI;
```